lsp/ts_ls.lua

```diff
@@ -78,4 +81,41 @@ return {
     'typescript.tsx',
   },
   root_markers = { 'tsconfig.json', 'jsconfig.json', 'package.json', '.git' },
+  on_attach = function(client)
+    -- ts_ls provides code actions that have a prefix `source.` that apply to the whole file. These are not discoverable
+    -- via the normal `vim.lsp.buf.code_action()` so add a command that allows access to them.
+    local function get_source_actions()
+      local codeActionKinds = client.server_capabilities.codeActionProvider.codeActionKinds
```

Comment:
Why aren't these discoverable via `code_action()` ? Can we enhance `code_action()` ?

Reply:
The typescript language server only returns these code actions if they're specified in `context.only`.

So there's nothing to fix on the nvim side. The command implementation actually does use `code_action()`.

Reply:
> The typescript language server only returns these code actions if they're specified in `context.only`.

But why? Is it because `code_action()` is sending position info, or diagnostics info? Do we need a way to tell `code_action()` not to send that info?

Reply:
That is really just how they've implemented it. You can see the code checks it [here](https://github.com/typescript-language-server/typescript-language-server/blob/0a034dae6c777733f4c2f6d5b151c5ef781fb662/src/lsp-server.ts#L779) where `kinds` is populated from [context.only](https://github.com/typescript-language-server/typescript-language-server/blob/0a034dae6c777733f4c2f6d5b151c5ef781fb662/src/lsp-server.ts#L769). The [refactor and quickfix actions](https://github.com/typescript-language-server/typescript-language-server/blob/0a034dae6c777733f4c2f6d5b151c5ef781fb662/src/lsp-server.ts#L770-L775) on the other hand have an extra `!kinds` check which means they're always returned if `context.only` is not provided.